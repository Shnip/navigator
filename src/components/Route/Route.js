import React from 'react';
import PropTypes from 'prop-types';
import { Draggable } from 'react-beautiful-dnd';
import { IoIosClose as Delete } from 'react-icons/io';


const Route = (props) => {
  const { route, index } = props;
  return (
    <Draggable draggableId={route.id} index={index}>
      {
        (provided, snapshot) => (
          <li className="routes__item list__item"
              ref={provided.innerRef}
              {...provided.draggableProps}
              {...provided.dragHandleProps} >
            <span className="list__body">{route.value}</span>
            <span data-delete={route.id}>
              <Delete className="routes__icon" size="1.5rem" />
            </span>
          </li>
        )
      }
    </Draggable>
  );
}

Route.propTypes = {
  route: PropTypes.object.isRequired
};

export default Route;
