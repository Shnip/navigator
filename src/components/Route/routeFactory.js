const uuid = require('uuidv4');

class Route {
  constructor(value, geoObject) {
    this.id = uuid()
    this.value = value
    this.geoObject = geoObject
  }
}

export default Route;
