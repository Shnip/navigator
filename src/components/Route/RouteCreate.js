import React from 'react';
import PropTypes from 'prop-types';
import Route from './routeFactory';
import {NotificationContainer, NotificationManager} from 'react-notifications';


class RouteCreate extends React.Component {
  state = {
    route: ''
  }

  componentDidMount() {
    this.createYaSearch()
  }

  createYaSearch = () => {
    const ymaps = window.ymaps;

    if (!ymaps) return false;

    ymaps.ready(() => {
      const suggestView = new ymaps.SuggestView(
        'route__input',
        {
            results: '5',
            provider: 'yandex#map',
        }
      );

      suggestView.events.add('select', (e) => {
        const selected = e.get('item').value;
        if (this.state.route !== selected) {
          this.setState({route: selected});
        }
      })
    })
  }

  onSubmit = async (e) => {
    e.preventDefault();

    let {route} = this.state;
    if (!route.trim()) return this.createNotification('error', 'Route is empty');

    let points = [];

    await this.geocode(route)
              .then(res => points = res.geoObjects.toArray())
              .catch(e => console.log(e))

    if (points.length === 0) return this.createNotification('error', 'Sorry, we can\'t find the address');

    const isCoordinates = route.match(/\d+\.?\d*,?\s?\d+\.?\d*/gi);
    if (isCoordinates) route = points[0].properties.get('text');

    const geoObject = Object.assign({},
      {coordinates: points[0].geometry.getCoordinates()},
      {balloon: points[0].properties.get('balloonContent')}
    );

    this.props.addRoute(new Route(route, geoObject));
    this.setState({route: ''});
  }

  geocode = (route) => {
    const ymaps = window.ymaps;
    if (!ymaps) return false

    return ymaps.geocode(route, {
      json: false,
      provider: 'yandex#map',
      results: 1
    })
  }

  createNotification = (type, message) => {
    switch (type) {
      case 'error':
        NotificationManager.error('Sorry, we can\'t find the address', 'Error', 6000);
        break;
      default:
        break;
    }
  }

  render() {
    return (
      <>
        <NotificationContainer/>
        <div className="routes__create">
          <h2 className="title title--white routes__title">
            <span className="text-bold">Create</span>
            &nbsp;
            <span className="text-thin">Route</span>
          </h2>
          <form name="routes__form" className="form routes__form" onSubmit={this.onSubmit}>
            <div className="form__input-group">
              <input type="text"
                     id="route__input"
                     className="form__input"
                     value={this.state.route}
                     placeholder="Москва, метро Арбатская или 55.753595 37.621031"
                     onChange={e => this.setState({route: e.target.value})}
                     />
              <button className="btn btn--light-theme">Add</button>
            </div>
          </form>
        </div>
      </>
    );
  }
}

RouteCreate.propTypes = {
  addRoute: PropTypes.func.isRequired
}

export default RouteCreate;
