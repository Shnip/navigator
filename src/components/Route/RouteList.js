import React from 'react';
import PropTypes from 'prop-types';
import { Droppable } from 'react-beautiful-dnd';
import Route from './Route';
import { FaBus, FaWalking, FaBicycle, FaCar } from 'react-icons/fa'
import classnames from 'classnames';

const RouteList = (props) => {
  const {routes, deleteRoute, handleMode, mode} = props;

  const handleDelete = (e) => {
    let target = e.target;
    const ul = e.currentTarget;

    while (target !== ul) {
      if (target.hasAttribute('data-delete')) {
        const id = target.dataset.delete;

        return deleteRoute(id)
      }

      target = target.parentNode;
    }
  }

  return (
    <section className="routes__list-wrapper">
      <h2 className="title routes__list-title">
        <span className="text-bold">Route</span>
        &nbsp;
        <span className="text-thin">List</span>
      </h2>

      <section className="routes__modes">
        <ul className="list route-modes" onClick={handleMode}>
          <li className={
            classnames("route-modes__mode",
                      {"route-modes__mode--active": mode === "auto"}
            )}
              data-mode="auto">
            <FaCar />
          </li>
          <li className={
            classnames("route-modes__mode",
                      {"route-modes__mode--active": mode === "bus"})}
              data-mode="bus">
            <FaBus />
          </li>
          <li className={
            classnames("route-modes__mode",
              {"route-modes__mode--active": mode === "pedestrian"})}
              data-mode="pedestrian">
            <FaWalking />
          </li>
          <li className={
            classnames("route-modes__mode",
                  {"route-modes__mode--active": mode === "bicycle"}
              )}
              data-mode="bicycle">
            <FaBicycle />
          </li>
        </ul>
      </section>

      {
        routes && routes.length > 0 ?
          <Droppable droppableId="1">
            {
              (provided, snapshot) => {
                return (
                  <ul className="list routes__list"
                      onClick={handleDelete}
                      ref={provided.innerRef}
                      style={{ backgroundColor: snapshot.isDraggingOver ? '#8bf1c4' : '' }}
                      {...provided.droppableProps}>
                    {
                      routes.map((route, index) => (
                        <Route route={route} index={index} key={route.id}/>
                      ))
                    }
                    {provided.placeholder}
                </ul>
                )
              }
            }
          </Droppable>
        :
        <span className="routes__empty">There are no routes yet</span>
      }
    </section>
  );
}

RouteList.propTypes = {
  routes: PropTypes.array,
  mode: PropTypes.string.isRequired,
  deleteRoute: PropTypes.func.isRequired,
  handleMode: PropTypes.func.isRequired,
};

export default RouteList;
