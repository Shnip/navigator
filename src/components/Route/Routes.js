import React from 'react';
import RouteCreate from './RouteCreate';
import RouteList from './RouteList';
import { DragDropContext } from 'react-beautiful-dnd';
import Map from '../Map/Map';


export default class Routes extends React.Component {
  state = {
    routes: [],
    mode: 'auto'
  }

  yaMap = null
  multiRoute = null

  addRoute = (route) => {
    const {routes} = this.state;

    if (routes.length > 0) {
      return this.setState({ routes: [...this.state.routes, route] }, this.updateMap.bind(this, true));
    }

    this.setState({ routes: [...this.state.routes, route] }, this.createMultiRoute);
  }

  deleteRoute = (id) => {
    const new_routes = this.state.routes.filter(route => route.id !== id);

    this.setState({ routes: new_routes }, this.updateMap);
  }

  onDragEnd = (result) => {
    if (!result.destination) {
      return;
    }

    const routes = this.reorder(
      this.state.routes,
      result.source.index,
      result.destination.index
    );

    this.setState({ routes }, this.updateMap)
  }

  reorder = (list, startIndex, endIndex) => {
    const result = Array.from(list);
    const [removed] = result.splice(startIndex, 1);
    result.splice(endIndex, 0, removed);

    return result;
  };

  createMultiRoute = () => {
    const {routes, mode} = this.state;
    const points = routes.map(route => route.geoObject.coordinates);
    const ymaps = window.ymaps;

    if (!ymaps) return false

    this.multiRoute = new ymaps.multiRouter.MultiRoute({
      referencePoints: points,
      params: {
        routingMode: mode
      }
    }, {
      boundsAutoApply: true
    });

    this.yaMap.geoObjects.add(this.multiRoute);
    this.setBalloon(points.length-1);
    this.multiRoute.editor.start();

    this.multiRoute.events.add("dragend", () => {
      this.centerMap();
    });
  }

  updateMap = (shouldsetBalloon = false) => {
    const points = this.state.routes.map(route => route.geoObject.coordinates);

    if(!this.multiRoute) return false

    this.multiRoute.model.setReferencePoints(points);

    if (points.length > 0) {
      this.multiRoute.events.once("boundschange", this.centerMap);
    }

    if (shouldsetBalloon) {
      this.setBalloon(points.length-1);
    }
  }

  setBalloon = (index) => {
    const ymaps = window.ymaps;

    this.multiRoute.model.events.once("requestsuccess", () => {
      const wayPoint = this.multiRoute.getWayPoints().get(index);

      ymaps.geoObject.addon.balloon.get(wayPoint);
      wayPoint.properties.set({
        balloonContentBody: this.state.routes[index].geoObject.balloon
      });
    });
  }

  setYaMap = (map) => {
    this.yaMap = map;
  }

  centerMap = () => {
    this.yaMap.setBounds(this.multiRoute.getBounds(), {
        checkZoomRange: true
    });
  }

  handleMode = (e) => {
    let target = e.target;
    const ul = e.currentTarget;
    while(target !== ul) {
      if (target.hasAttribute('data-mode')) {
        return this.setState({ mode: target.dataset.mode }, this.setRoutingModeToMap)
      }
      target = target.parentNode;
    }
  }

  setRoutingModeToMap = () => {
    if (!this.multiRoute) return false;

    this.multiRoute.model.setParams({
      routingMode: this.state.mode
    });
  }

  render() {
    return (
      <div className="container__content">
        <section className="routes">
          <RouteCreate addRoute={this.addRoute} />
          <DragDropContext onDragEnd={this.onDragEnd}>
            <RouteList  routes={this.state.routes}
                        deleteRoute={this.deleteRoute}
                        handleMode={this.handleMode}
                        mode={this.state.mode} />
          </DragDropContext>
        </section>
        <section className="map">
          <Map setYaMap={this.setYaMap} />
        </section>
      </div>
    );
  }
}
