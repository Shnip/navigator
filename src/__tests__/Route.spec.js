import React from 'react';
import { mount } from 'enzyme';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import Route from '../components/Route/Route';

describe('test Route component', () => {
  const onDragEndMock = jest.fn();

  const props = {
    route: {id: 1, value: 'Moscow'},
    index: 0
  }

  const component = mount(
    <DragDropContext onDragEnd={onDragEndMock}>
      <Droppable droppableId="1">
        {
          (provided, snapshot) =>
            <ul ref={provided.innerRef} {...provided.droppableProps}>
              <Route {...props} />
            </ul>
        }
      </Droppable>
    </DragDropContext>
  );

  it('renders Draggable', () => {
    expect(component.find('Draggable')).toHaveLength(1);
  });

  it('renders li', () => {
    expect(component.find('.routes__item')).toHaveLength(1);
  });

  it('renders Delete', () => {
    expect(component.find('IoIosClose')).toHaveLength(1);
  });

  it ('renders properly', () => {
    expect(component).toMatchSnapshot();
  });
})
