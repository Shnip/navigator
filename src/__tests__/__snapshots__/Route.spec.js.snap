// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Route component renders properly 1`] = `ReactWrapper {}`;
